### PR TITLE
track coverage and apply validation checks when testing discovery

### PR DESCRIPTION
**Pull request by mgoldey**
- checks yaml against expected discovery format, validates regex within yaml, checks for duplicate entity definitions, and checks for the existence of all expected files during testing
- adds coverage reports to testing in CI

### CODE REVIEW

File: launch.py [hunk 1]
```diff
@@ -52,13 +63,15 @@ class LaunchTarget(Enum):
...
 def pull_image(image):
-    client = docker.client.from_env()
+    """
+    Try to pull the most up to date version of this image.
+    If internet connectivity is down, don't error.
+    """
     try:
-        client.images.pull(image)
-    except:
+        get_docker_client().images.pull(image)
```
Comment (cbrochtrup) [inline launch.py:71]: This is much safer.

File: validate_yaml.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import os
+import sys
+
+import yaml
+from fire import Fire
+
+from validate_regex import validate_regex_in_yaml_file
+
+
+def print_or_throw_error(condition, success_message, failure_message, code=1):
```
Comment (cbrochtrup) [inline validate_yaml.py:17]: `Code` looks unused. Should it deleted or is this missing code?

File: validate_yaml.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                                  if new_elements.count(elem) > 1)
+        overlap = overlap.union(duplicated_elements)
+    failure = is_known or not is_unique
+    return (list(overlap if failure else unique_elements), not failure)
+
+
+def validate_file_exists(file_name):
+    "This wrapper checks if a file exists."
+    if not os.path.isfile(file_name):
+        throw_error("{} does not exist".format(file_name))
```
Comment (cbrochtrup) [inline validate_yaml.py:58]: We could be a little more specific.
```suggestion
        throw_error("{} file does not exist".format(file_name))
```

File: validate_yaml.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    return True
+
+
+def key_is_present(data, key):
```
Comment (cbrochtrup) [inline validate_yaml.py:62]: I think this function is not used. I don't see it with ctl-f.

**Review by cbrochtrup — APPROVED** [review]
This is wonderful with great documentation and doctests.

File: validate_yaml.py (hunk 2 above)
Comment (mgoldey) [reply validate_yaml.py:17]: Used in check_key. Cleaning up some other references to outdated function `throw_error`.

File: launch.py [hunk 5]
```diff
@@ -52,13 +63,15 @@ class LaunchTarget(Enum):
...
+    except requests.exceptions.HTTPError:
```
Comment (mgoldey) [inline launch.py:72]: I've just confirmed the docker error thrown here is a derivative class of this error. Namely, it's a docker.errors.APIError, which is derived from requests.exceptions.HTTPError, so we're good here.

File: launch.py [hunk 6]
```diff
@@ -30,6 +33,14 @@
 BUILTIN_NLPROCESSOR_MODELS = "builtin-nlprocessor-models"
 
 
+def get_docker_client():
+    """
+    Returns docker client.
+    Invoke this repeatedly instead of persisting an object in python
```
Comment (sv-github) [inline launch.py:39]: :+1: for docstrings

**Pull request merged**